### DISCUSSION TITLE
Future plans (or intentions) for Cosmopolitan's bare metal support?

### DISCUSSION THREAD

**Discussion by tkchia** [jart/cosmopolitan#564] [General]
I am curious to know more about any plans (or intentions) there might be, for Cosmopolitan's support for the bare metal platform.

(E.g. is there any plan to have it support console I/O through a VGA monitor and a PS/2 (or maybe USB?) keyboard, in addition to the serial UART?)

Thank you!

**Comment by jart** [comment]
We'd love to improve our bare metal support. We already have stdio support through the serial uart port. I also really want ethernet drivers to exist that can be linked-in automatically if you use the berkeley socket system calls. I'd also love to have VGA / PS2 / etc. support if the user explicitly uses libraries we'd provide for that functionality, but it's not something I'd bake into every binary.

**Reply by tkchia** [reply]
Hello @jart,

Thank you; let me see if I can propose and maybe prototype out a workable arrangement for the runtime library(s).

I think VGA support would be useful (for me at least) for testing Cosmopolitan on an actual bare-metal setup, since I do not have many x64 machines at hand that have RS-232 serial ports.

My current thinking is to implement a separate set of routines for IBM PC console I/O &mdash; perhaps modeled after the `<conio.h>` or `<graph.h>` routines in [Open Watcom](https://github.com/open-watcom/open-watcom-v2-wikidocs/blob/master/docs/clib.pdf) &mdash; then from there, see how to integrate it into Cosmopolitan's file descriptor machinery in a "zero-cost" way.

Thank you!

**Reply by jart** [reply]
What I'd recommend is any program where we want a VGA console, we'd say this in our main() file:

```c
STATIC_YOINK("vga_console");
```

I'd recommend we implement the font support using:

- https://justine.lol/NotoSansMono-Regular.ttf (we can check it in)
- https://github.com/jart/cosmopolitan/blob/master/third_party/stb/stb_truetype.c

Then we could update the `libc/calls/*-serial.c` code to also log to VGA console if it's linked.

**Reply by jart** [reply]
We already have some code in the repository which renders fonts to the terminal as block characters if you want to use it as a reference:

```
o//tool/viz/getglyph.com -v NotoSansMono-Regular.ttf | less
```

**Reply by tkchia** [reply]
Hello @jart,

Thanks for your suggestions!

I happened to be working on a quick prototype implementation of VGA output (https://github.com/jart/cosmopolitan/pull/588, https://github.com/tkchia/cosmopolitan/commit/c42c607c80c2920b585ee1a65a9c66ec68acd609).  For now, the code simply assumes that the legacy BIOS has set the screen to a text mode, and has programmed it with an appropriate font.

But I do hope to see how I can properly embed a Unicode font in a program, and use that.  I guess I now need to figure out whether it is better to render a TrueType font at runtime, or to precompute a bitmap font from it at build time. :thought_balloon: 

Thank you!

**Reply by jart** [reply]
Definitely don't hardcode bitmaps. Noto Sans Mono has 2919 glyphs. The file is 290kb. The STB library is tiny.

What you're doing now with the BIOS text mode sounds great too and also the tiniest. We could probably have it replicate serial uart to the BIOS screen display by default for all binaries.

**Reply by tkchia** [reply]
Hello @jart,

> What you're doing now with the BIOS text mode sounds great too and also the tiniest. We could probably have it replicate serial uart to the BIOS screen display by default for all binaries.

OK.  In that case, perhaps I will just implement a quick way to map UTF-8 multi-byte characters to CP437 characters where possible.

Thank you!

**Reply by jart** [reply]
Try the `unbing()` function which implements a multimapping of UNICODE -> CP437 for this purpose. https://github.com/jart/cosmopolitan/blob/master/libc/fmt/unbing.c For the inverse, see `kCp437`
